httpbakery/agent/agent.go

```diff
@@ -0,0 +1,161 @@
+agent cookie. It will most likely contain the data specified in the
+original agent cookie. This is used to determing that a successful
+status if for an agent login and not, for example, successfully
+accessing an interactive login page.
+*/
+package agent
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/loggo"
+	"gopkg.in/errgo.v1"
+
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+var logger = loggo.GetLogger("httpbakery.agent")
+
+// agentLogin defines the structure of an agent login cooke. It is also
+// returned in a successful agent login attempt to help indicate that an
+// agent login has occurred.
+type agentLogin struct {
+	Username  string            `json:"username"`
+	PublicKey *bakery.PublicKey `json:"public_key"`
+}
+
+// SetCookie creates a cookie in c which is suitable for performing agent
+// logins to u. The username will be combined with the public key taken
+// from c.Key.Public, c.Key must have been set before calling SetCookie.
+// If c is used with VisitWebPage interactions with u will be performed as
+// an agent.
+func SetCookie(c *httpbakery.Client, u *url.URL, username string) {
+	al := agentLogin{
+		Username:  username,
+		PublicKey: &c.Key.Public,
+	}
+	b, err := json.Marshal(al)
+	if err != nil {
+		// This shouldn't happen as the agentLogin type has to be marshalable.
+		panic(errgo.Notef(err, "cannot marshal cookie"))
+	}
+	v := base64.StdEncoding.EncodeToString(b)
+	c.Jar.SetCookies(u, []*http.Cookie{{
+		Name:  "agent-login",
+		Value: v,
+	}})
+}
+
+// VisitWebPage creates a function that can be used with
+// httpbakery.Client.VisitWebPage. The function uses c to access the
+// visit URL which should previously have been configured with SetCookie.
+func VisitWebPage(c *httpbakery.Client) func(u *url.URL) error {
+	return func(u *url.URL) error {
+		req, err := http.NewRequest("GET", u.String(), nil)
+		if err != nil {
+			return errgo.Notef(err, "cannot create request")
+		}
+		resp, err := c.Do(req)
+		if err != nil {
+			return errgo.Notef(err, "cannot perform request")
+		}
+		defer resp.Body.Close()
+		if err := getError(resp); err != nil {
+			return errgo.Notef(err, "cannot log in")
+		}
+		var al agentLogin
+		if err := unmarshalJSON(resp, &al); err != nil {
+			return errgo.Notef(err, "cannot unmarshal response")
+		}
+		if al.Username == "" || al.PublicKey == nil {
+			return errgo.Notef(err, "agent login failed")
+		}
+		return nil
+	}
+}
+
+// getError checks for an error in r. If the status code indicates there
+// is an error then getError attempts to unmarshal the error as an
+// httpbakery.Error. If it is unmarshalled succesfully then this error is
+// returned, otherwise an error based on the response status is returned.
+// getError returns nil only when the status code indicates there was no
+// error.
+func getError(r *http.Response) error {
+	if r.StatusCode < 400 {
```

Comment:
That's a good point, In practice it doesn't matter much as it will likely fail at the next hurdle, but I'll change it.